### PR TITLE
[CosmosDB] Add Sql stored procedure, udf and trigger cmdlets

### PR DESCRIPTION
**Pull request by MehaKaushik**
**History Notes:**  
(Fill in the following template if multiple notes are needed, otherwise PR title will be used for history note.)

[Component Name 1] (BREAKING CHANGE:) (az command:) make some customer-facing change.  
[Component Name 2] (BREAKING CHANGE:) (az command:) make some customer-facing change.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S165.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py [hunk 1]
```diff
@@ -319,6 +324,139 @@ def cli_cosmosdb_sql_container_update(client,
...
+                                       trigger_operation,
+                                       trigger_type):
+    if all(arg is None for arg in
+           [trigger_body, trigger_operation, trigger_type]):
+        return False
+
+    if trigger_body is not None:
+        sql_trigger_resource.body = trigger_body
+
+    if trigger_body is not None:
```
Comment (qianwens) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py]: >trigger_body [](start = 7, length = 12)

this should be "trigger_operation"

File: src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py [hunk 2]
```diff
@@ -319,6 +324,139 @@ def cli_cosmosdb_sql_container_update(client,
...
+        sql_trigger_resource.trigger_operation = trigger_operation
+
+    if trigger_body is not None:
```
Comment (qianwens) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py]: >trigger_body [](start = 7, length = 12)

trigger_type

Comment (MehaKaushik) [reply src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py]: Thanks for pointing this out. Changed it.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py (hunk 1 above)
Comment (MehaKaushik) [reply src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py]: Thanks for pointing this out. Changed it.

**Comment by MehaKaushik** [comment]
@zhoxing-ms @Juliehzl Can you please re-run the ci pipelines, the failing tests are not related to my changes. https://github.com/Azure/azure-cli/pull/11999/checks?check_run_id=420127845

**Comment by MehaKaushik** [comment]
Ping @zhoxing-ms  @Juliehzl

**Comment by yonzhan** [comment]
@qianwens  will help with this PR.

**Comment by Juliehzl** [comment]
> Ping @zhoxing-ms @Juliehzl

Done

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 3]
```diff
@@ -115,6 +115,32 @@ def load_arguments(self, _):
         c.argument('conflict_resolution_policy', options_list=['--conflict-resolution-policy', '-c'], type=shell_safe_json_parse, completer=FilesCompleter(), help='Conflict Resolution Policy, you can enter it as a string or as a file, e.g., --conflict-resolution-policy @policy-file.json or ' + SQL_GREMLIN_CONFLICT_RESOLUTION_POLICY_EXAMPLE)
         c.argument('throughput', help='The throughput of SQL container (RU/s). Default value is 400')
 
+# SQL stored procedure
+    with self.argument_context('cosmosdb sql stored_procedure') as c:
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: In CLI commands, we don't have commands with underline, but use `-`. In addition, may I know what does stored_procedure do?

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 4]
```diff
@@ -115,6 +115,32 @@ def load_arguments(self, _):
...
+        c.argument('account_name', options_list=['--account-name', '-a'], help="Cosmosdb account name", id_part=None)
+        c.argument('database_name', options_list=['--database-name', '-d'], help="Database name")
+        c.argument('container_name', options_list=['--container-name', '-c'], help="Container name")
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: These parameters are used several times. Please use CLIArgumentType to abstract them and reuse the argtype for your parameters. e.g.  https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/storage/_params.py#L48
https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/storage/_params.py#L120

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 5]
```diff
@@ -115,6 +115,32 @@ def load_arguments(self, _):
...
+        c.argument('stored_procedure_name', options_list=['--name', '-n'], help="StoredProcedure name")
+        c.argument('stored_procedure_body', options_list=['--body', '-b'], help="StoredProcedure body")
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: Can you make the help message more descriptive?vFor example, which kind of value is valid for storedprocedure body? The body value is string with any value or it should be in json format?

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 6]
```diff
@@ -115,6 +115,32 @@ def load_arguments(self, _):
...
+
+# SQL trigger
+    with self.argument_context('cosmosdb sql trigger') as c:
+        c.argument('account_name', options_list=['--account-name', '-a'], help="Cosmosdb account name", id_part=None)
+        c.argument('database_name', options_list=['--database-name', '-d'], help="Database name")
+        c.argument('container_name', options_list=['--container', '-c'], help="Container name")
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: The same as above.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 7]
```diff
@@ -115,6 +115,32 @@ def load_arguments(self, _):
...
+        c.argument('trigger_name', options_list=['--name', '-n'], help="Trigger name")
+        c.argument('trigger_body', options_list=['--body', '-b'], help="Trigger body")
+        c.argument('trigger_type', options_list=['--type', '-t'], help="Trigger type, values can be: Pre or Post")
+        c.argument('trigger_operation', options_list=['--operation'], help="The operation of the trigger. Values can be: All, Create, Update, Delete, Replace")
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: Please use get_enum_type() for allowed values.

Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: Please use get_enum_type() for allowed values.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 8]
```diff
@@ -115,6 +115,32 @@ def load_arguments(self, _):
...
+
+# SQL user defined function
+    with self.argument_context('cosmosdb sql user_defined_function') as c:
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: The same as above. underline is not allowed in azure cli. For user defined funcation, can we use `udf` in brief?

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 9]
```diff
@@ -115,6 +115,32 @@ def load_arguments(self, _):
...
+        c.argument('account_name', options_list=['--account-name', '-a'], help="Cosmosdb account name", id_part=None)
+        c.argument('database_name', options_list=['--database-name', '-d'], help="Database name")
+        c.argument('container_name', options_list=['--container', '-c'], help="Container name")
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: The same as above.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py [hunk 10]
```diff
@@ -78,6 +78,27 @@ def load_command_table(self, _):
         g.command('show', 'get_sql_container')
         g.command('delete', 'delete_sql_container')
 
+    with self.command_group('cosmosdb sql stored_procedure', cosmosdb_sql_sdk, client_factory=cf_sql_resources) as g:
+        g.custom_command('create', 'cli_cosmosdb_sql_stored_procedure_create_update')
+        g.custom_command('update', 'cli_cosmosdb_sql_stored_procedure_create_update')
+        g.command('list', 'list_sql_stored_procedures')
+        g.command('show', 'get_sql_stored_procedure')
+        g.command('delete', 'delete_sql_stored_procedure')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py]: Do you want to prompt confirmation message for delete command? It is recommended to add confirmation for delete command. e.g. https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/storage/commands.py#L82

File: src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py [hunk 11]
```diff
@@ -78,6 +78,27 @@ def load_command_table(self, _):
...
+
+    with self.command_group('cosmosdb sql trigger', cosmosdb_sql_sdk, client_factory=cf_sql_resources) as g:
+        g.custom_command('create', 'cli_cosmosdb_sql_trigger_create')
+        g.custom_command('update', 'cli_cosmosdb_sql_trigger_update')
+        g.command('list', 'list_sql_triggers')
+        g.command('show', 'get_sql_trigger')
+        g.command('delete', 'delete_sql_trigger')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py]: The same as above.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py [hunk 12]
```diff
@@ -78,6 +78,27 @@ def load_command_table(self, _):
...
+
+    with self.command_group('cosmosdb sql user_defined_function', cosmosdb_sql_sdk, client_factory=cf_sql_resources) as g:
+        g.custom_command('create', 'cli_cosmosdb_sql_user_defined_function_create_update')
+        g.custom_command('update', 'cli_cosmosdb_sql_user_defined_function_create_update')
+        g.command('list', 'list_sql_user_defined_functions')
+        g.command('show', 'get_sql_user_defined_function')
+        g.command('delete', 'delete_sql_user_defined_function')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py]: The same as above.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py [hunk 13]
```diff
@@ -320,6 +326,145 @@ def cli_cosmosdb_sql_container_update(client,
...
+
+    """Creates or Updates an Azure Cosmos DB SQL stored procedure """
+    sql_stored_procedure_resource = SqlStoredProcedureResource(id=stored_procedure_name)
+    sql_stored_procedure_resource.body = stored_procedure_body
+
+    sql_stored_procedure_create_update_resource = SqlStoredProcedureCreateUpdateParameters(
+        resource=sql_stored_procedure_resource,
+        options={})
+
+    return client.create_update_sql_stored_procedure(resource_group_name,
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py]: It looks this operation is a long running operation. We will recommend to support no wait for such operation and add an additional wait command to query operation status. Please refer to https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/network/commands.py#L351

File: src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py [hunk 14]
```diff
@@ -320,6 +326,145 @@ def cli_cosmosdb_sql_container_update(client,
...
+
+    """Creates or Updates an Azure Cosmos DB SQL user defined function """
+    sql_user_defined_function_resource = SqlUserDefinedFunctionResource(id=user_defined_function_name)
+    sql_user_defined_function_resource.body = user_defined_function_body
+
+    sql_user_defined_function_create_update_resource = SqlUserDefinedFunctionCreateUpdateParameters(
+        resource=sql_user_defined_function_resource,
+        options={})
+
+    return client.create_update_sql_user_defined_function(resource_group_name,
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py]: The same as above.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py [hunk 15]
```diff
@@ -320,6 +326,145 @@ def cli_cosmosdb_sql_container_update(client,
...
+    sql_trigger_resource = SqlTriggerResource(id=trigger_name)
+    sql_trigger_resource.body = trigger_body
+    sql_trigger_resource.trigger_type = trigger_type
+    sql_trigger_resource.trigger_operation = trigger_operation
+
+    sql_trigger_create_update_resource = SqlTriggerCreateUpdateParameters(
+        resource=sql_trigger_resource,
+        options={})
+
+    return client.create_update_sql_trigger(resource_group_name,
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py]: The same as above.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py (hunk 8 above)
Comment (MehaKaushik) [reply src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: We have used the complete name in portal and powershell cmdlets. To maintain consistency across platforms, would prefer to keep the full name.

**Comment by MehaKaushik** [comment]
@Juliehzl addressed your comments, please have a look.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 16]
```diff
@@ -115,28 +119,54 @@ def load_arguments(self, _):
...
+        c.argument('stored_procedure_body', options_list=['--body', '-b'], completer=FilesCompleter(), help="StoredProcedure body, you can enter it as a string or as a file, e.g., --body @sprocbody-file.json")
+
+# SQL trigger
+    with self.argument_context('cosmosdb sql trigger') as c:
+        c.argument('account_name', account_name_type, id_part=None)
+        c.argument('database_name', database_name_type)
+        c.argument('container_name', container_name_type)
+        c.argument('trigger_name', options_list=['--name', '-n'], help="Trigger name")
+        c.argument('trigger_body', options_list=['--body', '-b'], completer=FilesCompleter(), help="Trigger body, you can enter it as a string or as a file, e.g., --body @triggerbody-file.json")
+        c.argument('trigger_type', options_list=['--type', '-t'], arg_type=get_enum_type(TriggerType), help="Trigger type, values can be: Pre or Post")
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: `values can be: Pre or Post` should be removed.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py [hunk 17]
```diff
@@ -115,28 +119,54 @@ def load_arguments(self, _):
...
+        c.argument('trigger_operation', options_list=['--operation'], arg_type=get_enum_type(TriggerOperation), help="The operation of the trigger. Values can be: All, Create, Update, Delete, Replace")
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_params.py]: The same as above.

**Comment by qianwens** [comment]
@MehaKaushik , could you please resolve @Juliehzl 's comments, and let me know when it is ready to merge.

**Comment by MehaKaushik** [comment]
@Juliehzl thanks for pointing that out. Changed as suggested.

@qianwens Hey, please merge at your convenience.

**Comment by MehaKaushik** [comment]
The cloud pipeline failed with ERROR: Please ensure you have network connection. Error detail: HTTPSConnectionPool(host='anfclipython.blob.core.windows.net', port=443): Max retries exceeded with url: /cli-extensions/netappfiles_preview-0.3.2-py2.py3-none-any.whl (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fcf40866710>: Failed to establish a new connection: [Errno -2] Name or service not known',))


Please re-run the pipeline, the failing cause isnt related to my changes.

**Comment by yonzhan** [comment]
add to S166.

**Comment by MehaKaushik** [comment]
@qianwens Thanks for re-running it. They are successful now, please merge at your convenience.

**Pull request merged**